libgit/autogit_manager.go

```diff
@@ -0,0 +1,397 @@
+type AutogitManager struct {
+	config         libkbfs.Config
+	kbCtx          libkbfs.Context
+	kbfsInitParams *libkbfs.InitParams
+	log            logger.Logger
+	deferLog       logger.Logger
+	resetQueue     chan resetReq
+	queueDoneCh    chan struct{}
+	getNewConfig   getNewConfigFn
+
+	lock             sync.Mutex
+	resetsInQueue    map[string]resetReq // key: resetReq.dstPath()
+	resetsInProgress map[string]resetReq // key: resetReq.dstPath()
+}
+
+// NewAutogitManager constructs a new AutogitManager instance, and
+// launches `numWorkers` processing goroutines in the background.
+func NewAutogitManager(
+	config libkbfs.Config, kbCtx libkbfs.Context,
+	kbfsInitParams *libkbfs.InitParams, numWorkers int) *AutogitManager {
+	log := config.MakeLogger("")
+	am := &AutogitManager{
+		config:           config,
+		kbCtx:            kbCtx,
+		kbfsInitParams:   kbfsInitParams,
+		log:              log,
+		deferLog:         log.CloneWithAddedDepth(1),
+		resetQueue:       make(chan resetReq, 1000),
+		queueDoneCh:      make(chan struct{}),
+		resetsInQueue:    make(map[string]resetReq),
+		resetsInProgress: make(map[string]resetReq),
+	}
+	am.getNewConfig = am.getNewConfigDefault
+	go am.resetLoop(numWorkers)
+	return am
+}
+
+// Shutdown shuts down this manager.
+func (am *AutogitManager) Shutdown() {
+	close(am.resetQueue)
+	<-am.queueDoneCh
+}
+
+func (am *AutogitManager) getNewConfigDefault(ctx context.Context) (
+	context.Context, libkbfs.Config, string, error) {
+	return getNewConfig(ctx, am.config, am.kbCtx, am.kbfsInitParams, am.log)
+}
+
+func (am *AutogitManager) doReset(ctx context.Context, req resetReq) (
+	err error) {
+	am.log.CDebugf(ctx, "Processing reset request from %s/%s to %s/%s",
+		req.srcTLF.GetCanonicalPath(), req.srcRepo,
+		req.dstTLF.GetCanonicalPath(), req.dstDir)
+	defer func() {
+		am.deferLog.CDebugf(ctx, "Reset request completed: %+v", err)
+	}()
+
+	// Make a new single-op config for processing this request.
+	ctx, gitConfig, tempDir, err := am.getNewConfig(ctx)
+	if err != nil {
+		return err
+	}
+	// Cloned/pull data should be regular data blocks.
+	gitConfig.SetDefaultBlockType(keybase1.BlockType_DATA)
+	defer func() {
+		gitConfig.Shutdown(ctx)
+		rmErr := os.RemoveAll(tempDir)
+		if rmErr != nil {
+			am.log.CDebugf(
+				ctx, "Error cleaning storage dir %s: %+v\n", tempDir, rmErr)
+		}
+	}()
+
+	uniqID, err := makeUniqueID(ctx, gitConfig)
+	if err != nil {
+		return err
+	}
+
+	// Construct a src repo FS.
+	srcRepoFS, _, err := GetRepoAndID(
+		ctx, gitConfig, req.srcTLF, req.srcRepo, uniqID)
+	if err != nil {
+		return err
+	}
+
+	// And a dst parent checkout FS, which better already exist.
+	dstFS, err := libfs.NewFS(
+		ctx, gitConfig, req.dstTLF, req.dstDir, uniqID,
+		keybase1.MDPriorityNormal)
+	if err != nil {
+		return err
+	}
+
+	// Take the lock for the dst repo while resetting.
+	lockFile, err := dstFS.Create(autogitLockName(req.srcRepo))
```

Comment:
Yeah exactly, I didn't feel like making a way to specify the destination repo name -- right now it's always the same as the source repo name.  This lock does implicitly include the whole repo path (`dstPath`+`srcRepo`), since it includes the lock namespace for the `FS`.

Reply:
Sounds good to me! Would it ever happen that we are checking out repos from different TLFs into same `dstFS`, or is "checkout" of repos from a different TLF always be under a full path (e.g. `.kb_git/team/keybase/repo.git`)? If latter, then it seems we may never need a way to make `dstRepo` different from `srcRepo` in the future.

Reply:
Yeah, the caller should always do checkouts under a full, unique path.  Not enforced in this interface, but the next layer up will do it.